src/main/java/com/avaya/letscode/Offer.java

```diff
@@ -29,12 +29,13 @@ public default double getDiscountPercentage(Cart cart) {
 
     // For item level offers calculate item total based on offer.
     // This function will be implemented in all derived classes
-    public default int getDiscountedValue(Product product, int quantity) {
-        return product.getUnitPrice() * quantity;
+    public default int getDiscountedValue(CartItem cartItem) {
+        return cartItem.getProduct().getUnitPrice() * cartItem.getQuantityBeforeOffer();
     }
 
 
-    public default int getAdjustedQuantity(Product product, int quantity) {
-        return quantity;
+
+    public default int getAdjustedQuantity(CartItem cartItem) {
+        return cartItem.getQuantityBeforeOffer();
```

Comment:
All the methods in Offer interface (except getOfferName) should move to their respective, more specific interface - either CartOffer OR CartItemOffer. And must be removed from Offer interface.